2010-11-24  Richard Sandiford  <richard.sandiford@linaro.org>

	Launchpad #618684

	Backport from mainline:

	2010-04-10  Bernd Schmidt  <bernds@codesourcery.com>

	* reload1.c (eliminate_regs_in_insn): Don't restore an operand
	if doing so would replace the entire pattern.

=== modified file 'gcc/reload1.c'
--- old/gcc/reload1.c	2010-10-04 00:50:43 +0000
+++ new/gcc/reload1.c	2010-11-24 13:40:23 +0000
@@ -3567,7 +3567,10 @@
     {
       /* Restore the old body.  */
       for (i = 0; i < recog_data.n_operands; i++)
-	*recog_data.operand_loc[i] = orig_operand[i];
+	/* Restoring a top-level match_parallel would clobber the new_body
+	   we installed in the insn.  */
+	if (recog_data.operand_loc[i] != &PATTERN (insn))
+	  *recog_data.operand_loc[i] = orig_operand[i];
       for (i = 0; i < recog_data.n_dups; i++)
 	*recog_data.dup_loc[i] = orig_operand[(int) recog_data.dup_num[i]];
     }

